src/server/frontend_wayland/session_lock_v1.cpp

```diff
@@ -127,23 +172,32 @@ void mf::SessionLockManagerV1::Instance::lock(wl_resource* lock)
 
 // SessionLockV1
 
-mf::SessionLockV1::SessionLockV1(wl_resource* new_resource, mf::SessionLockManagerV1& manager)
+mf::SessionLockV1::SessionLockV1(wl_resource* new_resource, SessionLockManagerV1& manager)
     : mw::SessionLockV1(new_resource, Version<1>()),
       manager{manager},
-      lock_handle{manager.surface_stack->lock_screen()}
+      observer{std::make_shared<SessionLockV1Observer>(*this)}
 {
-    send_locked_event();
+    if (manager.try_lock(this))
+    {
+        send_locked_event();
+        manager.session_locker->register_interest(observer, manager.wayland_executor);
+    }
+    else
+        send_finished_event();
 }
 
 mf::SessionLockV1::~SessionLockV1()
 {
     if (client->is_being_destroyed())
     {
-        fatal_error("Screen locker died while screen was locked");
+        mir::log_warning("Screen locker died while screen was locked."
+                         "The compositor may be stuck in a locked state forever.");
+        manager.try_relinquish_locking_privilege(this);
     }
     else
     {
-        lock_handle->allow_to_be_dropped();
+        if (!manager.try_unlock(this))
+            mir::log_warning("Duplicate SessionLockV1 will not cause unlock");
     }
 }
```

Comment:
Oh! I was wondering about this, and it turns out this logic here is because we have a bug in our wrapper generator!

There are *two* requests that destroy `SessionLockV1`: `destroy()` and `unlock_and_destroy()`, but our wrapper generator doesn't expose that and instead just destroys the object in the implementation for both. This would be easier (and we'd be better able to be protocol-compliant) if the generator [did something different](https://github.com/canonical/mir/issues/3241)

Reply:
Probably worth fixing the generator and then this

Reply:
I can work on that today :+1: